### PR TITLE
Clean up and harden various methods around trait substs

### PR DESCRIPTION
**Pull request by oli-obk**
r? @lcnr

### CODE REVIEW

File: compiler/rustc_trait_selection/src/infer.rs [hunk 1]
```diff
@@ -113,12 +113,17 @@ impl<'tcx> InferCtxtExt<'tcx> for InferCtxt<'tcx> {
...
     ) -> traits::EvaluationResult {
-        let trait_ref =
-            ty::TraitRef { def_id: trait_def_id, substs: self.tcx.mk_substs_trait(ty, params) };
+        let trait_ref = self.tcx.mk_trait_ref(trait_def_id, self_ty, params);
+
+        debug_assert_eq!(
+            self.tcx.generics_of(trait_def_id).count() - 1,
+            params.len(),
+            "wrong number of generic parameters for {trait_def_id:?}, did you accidentally include the self-type in the params list?"
+        );
```
Comment (lcnr) [inline compiler/rustc_trait_selection/src/infer.rs:126]: isn't that already checked by `mk_trait_ref`?

**Comment by oli-obk** [comment]
@rustbot ready

cc @lcnr

File: src/tools/clippy/clippy_lints/src/methods/unnecessary_to_owned.rs [hunk 2]
```diff
@@ -490,7 +490,7 @@ fn is_to_string_on_string_like<'a>(
         && let GenericArgKind::Type(ty) = generic_arg.unpack()
         && let Some(deref_trait_id) = cx.tcx.get_diagnostic_item(sym::Deref)
         && let Some(as_ref_trait_id) = cx.tcx.get_diagnostic_item(sym::AsRef)
-        && (implements_trait(cx, ty, deref_trait_id, &[cx.tcx.types.str_.into()]) ||
+        && (get_associated_type(cx, ty, deref_trait_id, "Target") == Some(cx.tcx.types.str_) ||
```
Comment (lcnr) [inline src/tools/clippy/clippy_lints/src/methods/unnecessary_to_owned.rs:493]: nice to see that there wasn't ever a test for `Deref<Target = NotString>` :grin:

**Pull request merged**